### PR TITLE
[Model] Support DeepSeek-OCR-2

### PR DESCRIPTION
**Pull request by LiuLi1998**
## Purpose
Support New Model: DeepSeek-OCR-2

## Test Result

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/transformers_utils/processors/deepseek_ocr2.py
Comment (Isotr0py) [inline vllm/transformers_utils/processors/deepseek_ocr2.py:1]: I think Deepseek-OCR2 uses same processor like Deepseek-OCR?

File: vllm/model_executor/models/deepseek_ocr2.py
Comment (Isotr0py) [inline vllm/model_executor/models/deepseek_ocr2.py:1]: According to Deepseek-OCR2's technical report, I think we just need to replace CLIP with Qwen2Encoder at `deepseek_ocr.py`.

File: vllm/transformers_utils/processors/deepseek_ocr2.py
Comment (LiuLi1998) [reply vllm/transformers_utils/processors/deepseek_ocr2.py:1]: Slightly different in tokenize_with_images, and different image size limit

File: vllm/model_executor/models/deepseek_ocr2.py
Comment (LiuLi1998) [reply vllm/model_executor/models/deepseek_ocr2.py:1]: Thx for review. I reformat the code, remove some redundant code

**Comment by LiuLi1998** [comment]
> @LiuLi1998 It seems there is an issue with accuracy, [request by doc](https://docs.vllm.ai/projects/recipes/en/latest/DeepSeek/DeepSeek-OCR.html#offline-ocr-tasks)
> 
> ```
> Response costs: 6.34s
> Generated text: 17,000
> CASHIER: 17,000
> CASHIER ID: 17,000
> CASHIER NAME: 17,000
> CASHIER TOTALS: 17,000
> 
> CASHIER ID: 17,000
> CASHIER NAME: 17,000
> CASHIER TOTALS: 17.000
> 
> CASHIER ID: 17,000
> CASHIER NAME: 17,000
> CASHIER TOTALS: CASHIER TOTALS: 17,000
> 
> CASHIER ID: 17,000
> CASHIER NAME: 17,000
> CASHIER TOTALS: 17,000
> 
> CASHIER ID: 17,000
> CASHIER NAME: 17,001
> CASHIER TOTALS: 17,000
> 
> CASHIER ID: 17,000
> CASHIER NAME: 18,000
> CASHIER TOTALS: 18,000
> 
> CASHIER ID: 18,000
> CASHIER NAME: 18,001
> CASHIER TOTALS: 18,000
> 
> CASHIER ID: 18,000
> CASHIER NAME: 19,000
> CASHIER TOTALS: 19,000
> 
> CASHIER ID: 19,000
> CASHIER NAME: 19,001
> CASHIER TOTALS: 19,000
> 
> CASHIER ID: 19,000
> CASHIER NAME: 20,000
> CASHIER TOTALS: 20,000
> 
> CASHIER ID: 20,000
> CASHIER NAME: 20,001
> CASHIER TOTALS: 20,000
> 
> CASHIER ID: 20,000
> CASHIER NAME: 21,000
> CASHIER TOTALS: 21,000
> 
> CASHIER ID: 21,000
> CASHIER NAME: 21,001
> CASHIER TOTALS: 21,000
> 
> CASHIER ID: 21,000
> CASHIER NAME: 22,000
> CASHIER TOTALS: 22,000
> 
> CASHIER ID: 22,000
> CASHIER NAME: 22,001
> CASHIER TOTALS: 22,000
> 
> CASHIER ID: 22,000
> CASHIER NAME: 23,000
> CASHIER TOTALS: 23,000
> 
> CASHIER ID: 23,000
> CASHIER NAME: 23,001
> CASHIER TOTALS: 23,000
> 
> CASHIER ID: 23,000
> CASHIER NAME: 24,000
> CASHIER TOTALS: 24,000
> 
> CASHIER ID: 24,000
> CASHIER NAME: 24,001
> CASHIER TOTALS: 24,000
> 
> CASHIER ID: 24,000
> CASHIER NAME: 25,000
> CASHIER TOTALS: 25,000
> 
> CASHIER ID: 25,000
> CASHIER NAME: 25,001
> CASHIER TOTALS: 25,000
> 
> CASHIER ID: 25,000
> CASHIER NAME: 26,000
> CASHIER TOTALS: 26,000
> 
> CASHIER ID: 26,000
> CASHIER NAME: 26,001
> CASHIER TOTALS: 26,000
> 
> CASHIER ID: 26,000
> CASHIER NAME: 27,000
> CASHIER TOTALS: 27,000
> 
> CASHIER ID: 27,000
> CASHIER NAME: 27,001
> CASHIER TOTALS: 27,000
> 
> CASHIER ID: 27,000
> CASHIER NAME: 28,000
> CASHIER TOTALS: 28,000
> 
> CASHIER ID: 28,000
> CASHIER NAME: 28,001
> CASHIER TOTALS: 28,000
> 
> CASHIER ID: 28,000
> CASHIER NAME: 29,000
> CASHIER TOTALS: 29,000
> 
> CASHIER ID: 29,000
> CASHIER NAME: 29,001
> CASHIER TOTALS: 29,000
> 
> CASHIER ID: 29,000
> CASHIER NAME: 30,000
> CASHIER TOTALS: 30,000
> 
> CASHIER ID: 30,000
> CASHIER NAME: 30,001
> CASHIER TOTALS: 30,000
> 
> CASHIER ID: 30,000
> CASHIER NAME: 31,000
> CASHIER TOTALS: 31,000
> 
> CASHIER ID: 31,000
> CASHIER NAME: 31,001
> CASHIER TOTALS: 31,000
> 
> CASHIER ID: 31,000
> CASHIER NAME: 32,000
> CASHIER TOTALS: 32,000
> 
> CASHIER ID: 32,000
> CASHIER NAME: 32,001
> CASHIER TOTALS: 32,000
> 
> CASHIER ID: 32,000
> CASHIER NAME: 33,000
> CASHIER TOTALS: 33,000
> 
> CASHIER ID: 33,000
> CASHIER NAME: 33,001
> CASHIER TOTALS: 33,000
> 
> CASHIER ID: 33,000
> CASHIER NAME: 34,000
> CASHIER TOTALS: 34,000
> 
> CASHIER ID: 34,000
> CASHIER NAME: 34,001
> CASHIER TOTALS: 34,000
> 
> CASHIER ID: 34,000
> CASHIER NAME: 35,000
> CASHIER TOTALS: 35,000
> 
> CASHIER ID: 35,000
> CASHIER NAME: 35,001
> CASHIER TOTALS: 35,000
> 
> CASHIER ID: 35,000
> CASHIER NAME: 36,000
> CASHIER TOTALS: 36,000
> 
> CASHIER ID: 36,000
> CASHIER NAME: 36,001
> CASHIER TOTALS: 36,000
> 
> CASHIER ID: 36,000
> CASHIER NAME: 37,000
> CASHIER TOTALS: 37,000
> 
> CASHIER ID: 37,000
> CASHIER NAME: 37,001
> CASHIER TOTALS: 37,000
> 
> CASHIER ID: 37,000
> CASHIER NAME: 38,000
> CASHIER TOTALS: 38,000
> 
> CASHIER ID: 38,000
> CASHIER NAME: 38,001
> CASHIER TOTALS: 38,000
> 
> CASHIER ID: 38,000
> CASHIER NAME: 39,000
> CASHIER TOTALS: 39,000
> 
> CASHIER ID: 39,000
> CASHIER NAME: 39,001
> CASHIER TOTALS: 39,000
> 
> CASHIER ID: 39,000
> CASHIER NAME: 40,000
> CASHIER TOTALS: 40,000
> 
> CASHIER ID: 40,000
> CASHIER NAME: 40,001
> CASHIER TOTALS: 40,000
> 
> CASHIER ID: 40,000
> CASHIER NAME: 41,000
> CASHIER TOTALS: 41,000
> 
> CASHIER ID: 41,000
> CASHIER NAME: 41,001
> CASHIER TOTALS: 41,000
> 
> CASHIER ID: 41,000
> CASHIER NAME: 42,000
> CASHIER TOTALS: 42,000
> 
> CASHIER ID: 42,000
> CASHIER NAME: 42,001
> CASHIER TOTALS: 42,000
> 
> CASHIER ID: 42,000
> CASHIER NAME: 43,000
> CASHIER TOTALS: 43,000
> 
> CASHIER ID: 43,000
> CASHIER NAME: 43,001
> CASHIER TOTALS: 43,000
> 
> CASHIER ID: 43,000
> CASHIER NAME: 44,000
> CASHIER TOTALS: 44,000
> 
> CASHIER ID: 44,000
> CASHIER NAME: 44,001
> CASHIER TOTALS: 44,000
> 
> CASHIER ID: 44,000
> CASHIER NAME: 45,000
> CASHIER TOTALS: 45,000
> 
> CASHIER ID: 45,000
> CASHIER NAME: 45,001
> CASHIER TOTALS: 45,000
> 
> CASHIER ID: 45,000
> CASHIER NAME: 46,000
> CASHIER TOTALS: 46,000
> 
> CASHIER ID: 46,000
> CASHIER NAME: 46,001
> CASHIER TOTALS: 46,000
> 
> CASHIER ID: 46,000
> CASH
> ```

Working on it

File: vllm/model_executor/models/deepseek_ocr2.py [hunk 1]
```diff
@@ -0,0 +1,569 @@
...
+        local_features = self.sam_model(patches)
+        local_features = self.vision_model(local_features)
+
+        features = self.projector(local_features)
+
+        _, _, dim = features.shape
+
+        return features.view(-1, dim)
+
+    def _pixel_values_to_embedding(
```
Comment (RWayne93) [inline vllm/model_executor/models/deepseek_ocr2.py:354]: adjusting the pixel values function here to match the deepseek huggingface repo implementation fixed some accuracy and repetitiveness i was seeing from the model.

```python
    def _pixel_values_to_embedding(
        self,
        pixel_values: torch.Tensor,
        images_crop: torch.Tensor,
        images_spatial_crop: torch.Tensor,
    ) -> NestedTensors:
        images_in_this_batch = []

        is_tiled = (images_spatial_crop[:, 0] > 1) | (images_spatial_crop[:, 1] > 1)
        patches_per_image = torch.where(is_tiled, images_spatial_crop.prod(dim=-1), 0)
        images_crop = images_crop.split(patches_per_image.tolist())
        for jdx in range(images_spatial_crop.size(0)):
            patches = images_crop[jdx]
            image_ori = pixel_values[[jdx]]

            global_features = self._encode_global_features(image_ori)
            local_features = self._encode_local_features(patches)

            if local_features is not None:
                if self.global_view_pos == "head":
                    # Put global tokens first when requested by config.
                    combined = torch.cat(
                        [
                            global_features,
                            local_features,
                            self.view_seperator[None, :],
                        ],
                        dim=0,
                    )
                else:
                    # Default ordering from HF implementation.
                    combined = torch.cat(
                        [local_features, global_features, self.view_seperator[None, :]],
                        dim=0,
                    )
            else:
                combined = torch.cat(
                    [global_features, self.view_seperator[None, :]], dim=0
                )

            images_in_this_batch.append(combined)

        return images_in_this_batch
```

Comment (leihuang-sketch) [reply vllm/model_executor/models/deepseek_ocr2.py:354]: I still obtained duplicate and inaccurate results with your code.
```
Response costs: 9.52s
Generated text: 17,000
CASHIER: 17,000
CASHIER ID: 17,000
CASHIER NAME: 17,000
CASHIER TOTALS: 17,000

CASHIER ID: 17,000
CASHIER NAME: 17,000

CASHIER TOTALS: 17,000

CASHIER ID: 17.000
CASHIER NAME: 17,000
CASHIER TOTALS: CASHIER ID: 17,000

CASHIER TOTALS: 17,000

CINNAMON SUGAR
1
GRAND TOTAL
17,000

CASHIER TOTALS: 17,000

CASHIER TOTALS: 17,000

CASHIER ID: 17,000
CASHIER TOTALS: 17,000

CASHIER TOTALS: CASHIER ID: 17,000

CASHIER TOTALS: CASHIER TOTALS: CASHIER TOTALS: CASHIER TOTALS:
```

**Comment by LiuLi1998** [comment]
> @LiuLi1998 @RWayne93 Downgrading the Transformers version(4.51.3) can temporarily solve this problem. However, this is not the fundamental reason. The root cause lies in the different implementations of `qwen2_decoder` in different versions of transformers. There is an accuracy issue with `deepseek_encoder2` on version `4.57.6`, which does not exist on version `4.51.3` The `deepseek_encoder2` requires updating the `causal_mask`, but the relevant method has been removed in version `4.57.6`, resulting in the inability to update the `causal_mask` according to the logic of the `deepseek_encoder2`

I can try to refactor this part of the logic to align with the newer Transformers API while preserving the necessary causal mask behavior.
Thanks for the detailed analysis!

**Comment by leihuang-sketch** [comment]
@LiuLi1998 Can you grant me permission? I have already completed the code and verification. Thank you

**Comment by leihuang-sketch** [comment]
@LiuLi1998 I identify the mask injection function to determine the version, and inject masks through different logic for new and old versions

**Comment by LiuLi1998** [comment]
@leihuang-sketch @Isotr0py Would you mind reviewing this PR? accuracy issue is solved

**Comment by LiuLi1998** [comment]
> @LiuLi1998 Can you grant me permission? I have already completed the code and verification. Thank you

you already have, 
image

File: vllm/model_executor/models/deepseek_ocr2.py (hunk 1 above)
Comment (LiuLi1998) [reply vllm/model_executor/models/deepseek_ocr2.py:354]: solved in deepencoder2.py, thx for @leihuang-sketch

**Comment by RWayne93** [comment]
> > @LiuLi1998 @RWayne93 Downgrading the Transformers version(4.51.3) can temporarily solve this problem. However, this is not the fundamental reason. The root cause lies in the different implementations of `qwen2_decoder` in different versions of transformers. There is an accuracy issue with `deepseek_encoder2` on version `4.57.6`, which does not exist on version `4.51.3` The `deepseek_encoder2` requires updating the `causal_mask`, but the relevant method has been removed in version `4.57.6`, resulting in the inability to update the `causal_mask` according to the logic of the `deepseek_encoder2`
> 
> I can try to refactor this part of the logic to align with the newer Transformers API while preserving the necessary causal mask behavior. Thanks for the detailed analysis!

good catch. I didn't check the Transformers version in my initial attempts. I assumed it was on the most recent version when i was checking for accuracy based on the deepseek hf repo examples. hopefully this gets merged soon.

**Review by Isotr0py — APPROVED** [review]
LGTM now

File: vllm/model_executor/models/deepencoder2.py [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+    def __init__(
+        self,
+        decoder_layer: int = 24,
+        max_position_embeddings: int = 131072,
+        hidden_dimension: int = 896,
+        num_attention_heads: int = 14,
+        num_key_value_heads: int = 2,
+        intermediate_size: int = 4864,
+        vocab_size: int = 151936,
+        attn_implementation: str = "sdpa",  # ⭐
```
Comment (ywang96) [inline vllm/model_executor/models/deepencoder2.py:34]: Can we remove in-line emoji's? (and the rest of them)

Comment (ywang96) [inline vllm/model_executor/models/deepencoder2.py:34]: I'm also wondering if we should port this model to vLLM since `sdpa` could be really slow. I'm curious if you have measure the performance difference between sdpa and flash attention. If we use `transformers` directly there's no easy for us to use flash attention since we have our own fork.

**Review by ywang96 — COMMENTED** [review]
Thanks for the contribution! Other than the reviews above I have two final comments

File: vllm/model_executor/models/deepencoder2.py (hunk 2 above)
Comment (LiuLi1998) [reply vllm/model_executor/models/deepencoder2.py:34]: Thxfor the review. Since the PR has already been merged, should I open another PR to remove these comments?

Comment (LiuLi1998) [reply vllm/model_executor/models/deepencoder2.py:34]: I haven’t run performance tests yet, but I’ll try porting it later and see how it goes.
